store/tikv/mock-tikv/mvcc_leveldb.go

```diff
@@ -0,0 +1,724 @@
+func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
+	mvcc.mu.RLock()
+	defer mvcc.mu.RUnlock()
+
+	iter, currKey, err := newScanIterator(mvcc.db, startKey, endKey)
+	defer iter.Release()
+	if err != nil {
+		log.Error("scan new iterator fail:", errors.ErrorStack(err))
+		return nil
+	}
+
+	ok := true
+	var pairs []Pair
+	for len(pairs) < limit && ok {
+		value, err := getValue(iter, currKey, startTS, isoLevel)
+		if err != nil {
+			pairs = append(pairs, Pair{
+				Key: currKey,
+				Err: errors.Trace(err),
+			})
+		}
+		if value != nil {
+			pairs = append(pairs, Pair{
+				Key:   currKey,
+				Value: value,
+			})
+		}
+
+		skip := skipDecoder{currKey}
+		ok, err = skip.Decode(iter)
+		if err != nil {
+			log.Error("seek to next key error:", errors.ErrorStack(err))
+			break
+		}
+		currKey = skip.currKey
+	}
+	return pairs
+}
+
+// ReverseScan implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) ReverseScan(startKey, endKey []byte, limit int, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
+	return nil
+}
+
+// Prewrite implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) Prewrite(mutations []*kvrpcpb.Mutation, primary []byte, startTS uint64, ttl uint64) []error {
+	mvcc.mu.Lock()
+	defer mvcc.mu.Unlock()
+
+	anyError := false
+	errs := make([]error, 0, len(mutations))
+	for _, m := range mutations {
+		err := prewriteMutation(mvcc.db, m, startTS, primary, ttl)
+		errs = append(errs, err)
+		if err != nil {
+			anyError = true
+		}
+	}
+	if anyError {
+		return errs
+	}
+
+	return errs
+}
+
+func prewriteMutation(db *leveldb.DB, mutation *kvrpcpb.Mutation, startTS uint64, primary []byte, ttl uint64) error {
+	startKey := mvccEncode(mutation.Key, lockVer)
+	iter := newIterator(db, &util.Range{
+		Start: startKey,
+	})
+	defer iter.Release()
+
+	dec := lockDecoder{
+		expectKey: mutation.Key,
+	}
+	ok, err := dec.Decode(iter)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if ok {
+		if dec.lock.startTS != startTS {
+			return dec.lock.lockErr(mutation.Key)
+		}
+		return nil
+	}
+
+	dec1 := valueDecoder{
+		expectKey: mutation.Key,
+	}
+	ok, err = dec1.Decode(iter)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if ok && dec1.value.valueType != typeRollback && dec1.value.commitTS >= startTS {
+		return ErrRetryable("write conflict")
+	}
+
+	lock := mvccLock{
+		startTS: startTS,
+		primary: primary,
+		value:   mutation.Value,
+		op:      mutation.GetOp(),
+		ttl:     ttl,
+	}
+	writeKey := mvccEncode(mutation.Key, lockVer)
+	writeValue, err := lock.MarshalBinary()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Put(writeKey, writeValue, nil)
+	return errors.Trace(err)
+}
+
+// Commit implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) Commit(keys [][]byte, startTS, commitTS uint64) error {
+	mvcc.mu.Lock()
+	defer mvcc.mu.Unlock()
+
+	for _, k := range keys {
+		err := commitKey(mvcc.db, k, startTS, commitTS)
```

Comment:
Like prewrite, tikv's commit first generates write batch, then commit the write batch if no error happens.